Create CWL to run mpileup and filter output

tncowart opened the pull request:
[Create CWL to run mpileup and filter output](https://app.asana.com/0/1202519196726700/1202533373195957)

Turns

```shell
NAME=$(head -${SLURM_ARRAY_TASK_ID} scripts/mpileup_variables.txt | tail -1 | awk '{print $1}')
ALIGNMENTS=$(head -${SLURM_ARRAY_TASK_ID} scripts/mpileup_variables.txt | tail -1 | awk '{print $2}')

samtools mpileup \
        -d 0 \
        -f /data/reddylab/Reference_Data/Genomes/hg38/hg38.fa \
        -l /data/reddylab/gjohnson/scz/genotype_information/scz_variant_positions_from_genotyping_hg38.txt \
        ${ALIGNMENTS} \
| awk '{if($3!="N" && $3!="n") print $0}' \
| gzip > count_files/mpileups/${NAME}_mpileups.txt.gz
```

Into CWL workflows. The full pipeline is in 01_mpileups, with each step in a separate subdirectory. 
* `samtools mpileup` is in `01_mpileup/samtools_mpileup`
* `awk  '{if($3!="N" && $3!="n") print $0}'` is in `01_mpileup/filter_unknown_ref_base`
* `gzip` is in `gzip`

The workflow can be run against a single name/alignment pair by using `01_mpileups/pileups_workflow.cwl` directory, or against many pairs at once using `bird_workflow.cwl`, which will "scatter" the work -- `sample_bird_workflow.yml` shows how to set the input values up to be scattered.

tncowart committed:
CWLs/Workflow for pipeline steps 2 & 3 of gen and filter mpileups

tncowart committed:
Add samtools cwl and sample yml

The sample file paths are hardcoded for HARDAC

tncowart committed:
Incorporate pileup filteringinto overall BIRD workflow

* Also creates overall BIRD workflow

tncowart committed:
Add some new lines to the ends of files

alexbarrera commented on 01_mpileups/filter_unknown_ref_base/filter_unknown_ref_base.cwl:
```diff
@@ -0,0 +1,11 @@
+cwlVersion: v1.2
+class: CommandLineTool
+baseCommand: awk
+arguments: ['{if($3!="N" && $3!="n") print $0}']
```
I guess is pure style, but that statement could be a bit more succinct:
```
arguments: ['($3!="N" && $3!="n")']
```
That said, probably better to keep sugar syntax out of the equation, since the lab members using this code in the future will need to translate a lot of code from their scripts to the newly assembled pipeline.

alexbarrera commented on 01_mpileups/filter_unknown_ref_base/filter_unknown_ref_base.cwl:
```diff
@@ -0,0 +1,11 @@
+cwlVersion: v1.2
+class: CommandLineTool
```
I'm missing `hints` that make CWL code more flexible. Commenting on this CommandLineTool but it applies to the others. The idea being having a minimal base image with basic unix commands. Something like:
```
hints:
  DockerRequirement:
    dockerPull: "ubuntu:xenial"
```
Or similar

alexbarrera commented on 01_mpileups/pileups_workflow.cwl:
```diff
@@ -0,0 +1,35 @@
+inputs:
+  mpileup_depth: int
+  genome_file: File
+  locations: File
+  alignments: File
```
Can we keep the names of the variables consistent in nested subworkflows? For instance here `locations` is the same `mpileup_locations` in `bird_workflow.cwl`. Also I prefer using singular and plural variable names instead of suffices like `_array`, `_list` and such.

alexbarrera commented on 01_mpileups/pileups_workflow.cwl:
```diff
@@ -0,0 +1,35 @@
+cwlVersion: v1.2
```
Is there any reason why not to scatter over this subworkflow? That way if in the future you just want to run `pipeline_workflow.cwl` over multiple files you should be able to do so.

tncowart replied:
I do scatter "over this workflow" -- at the higher level bird_workflow.cwl. If you want to run just pipeline_workflow.cwl over a series of files, you would write pipeline_workflow_scatter.cwl that does the scattering. 

I've basically applied a common functional programming idiom here. If we have a function `f` that accepts a single value, and we want to apply it to a list of values, we use a higher level function, like `map(f, list_of_values)`. This "feels right" to me -- separating the function itself from the method of iteration. 

Additionally, adding the scattering into pipeline_workflow.cwl would hurt clarity.

But also, I don't think you can do the thing you're talking about -- scattering applies to a single workflow step and if you need to scatter over multiple steps you put them into a subworkflow, which is what I did. Looking at the spec, I don't see a different way to apply a scatter over multiple steps.

alexbarrera replied:
I got confused, what you are doing here is what I have been doing as well, which does in fact work well with either one or multiple files

alexbarrera commented on 01_mpileups/sample_pileups_workflow.yml:
```diff
@@ -0,0 +1,14 @@
+sample_pileup:
+  class: File
+  path: "filter_unknown_ref_base/sample_pileup.txt"
+name: "test"
```
I created a chr20 only set of files that I wonder if we want to put somewhere for testing and debugging proposes. Here is the `yaml` I used to test the workflow:
```yaml
name_array:
  - Input1
  - Output1
alignments_array:
  - { class: File, path: "/data/reddylab/Alex/tmp/bird_workflow_test_dir/Input1_lengthFiltered.chr20.bam" }
  - { class: File, path: "/data/reddylab/Alex/tmp/bird_workflow_test_dir/Output1.chr20.bam" }
mpileup_depth: 0
genome_file:
  class: File
  path: "/data/reddylab/Reference_Data/Genomes/hg38/hg38.fa"
mpileup_locations:
  class: File
  path: "/data/reddylab/Alex/tmp/bird_workflow_test_dir/scz_variant_positions_from_genotyping_hg38.chr20.txt"

```

tncowart replied:
Sure. In reddylab there are a few directories already that might be good places to store the test data -- `/data/reddylab/Datasets`, `/data/reddylab/TestData`, or `/data/reddylab/projects`

alexbarrera commented on 01_mpileups/samtools_mpileup/samtools_mpileup.cwl:
```diff
@@ -0,0 +1,28 @@
+cwlVersion: v1.2
```
Same as above in terms of adding a hint for a DockerRequirement. Luckily, you can reuse simply this image already available:
```cwl
 hints:
    DockerRequirement:
      dockerPull: dukegcb/samtools:1.3
```

alexbarrera reviewed (changes requested):
This is looking good, thanks Thomas. I have scattered a few comments across the files as I saw fit (in retrospect, I should have saved many of those for the review, but hey..).

I few comments:
- Having `DockerRequirements` as `hint` in `CommandLineTool`s I think is important in order to ensure maximum compatibility. In this case you would need to make a companion Docker image for a minimal base linux image.
- I like what you did of including example `yaml`s all across! If some of those folders get too busy we can always break them up into a dedicated example `yaml`s folder
- For the above point to be more convenient, I wonder if we can share a folder in Box with test-able sized files. CWL should be able to use URIs for the files (if we want to be fancy), but at a minimum it would be good for others to test and run the code *fast*.

tncowart committed:
Address PR change requests

* Add docker hints
* normalize input/output names

tncowart committed:
Remove compression step

I don't think we don't need to compress the output file here,
as it's going to get piped right into a python script (that's another PR/branch).
It may make sense to compress the result of the python script, though

alexbarrera commented on 01_mpileups/gzip/gzip.cwl:
```diff
@@ -1,12 +1,16 @@
+
+hints:
+    DockerRequirement:
+      dockerPull: alpine@sha256:06b5d462c92fc39303e6363c65e074559f8d6b1363250027ed5053557e3398c5
```
Any chance we can use a tagged image (for example `alpine:20220715`).

tncowart replied:
If you want. The image associated with a tag can change, which can be undesirable. In this case, it's fine, but it's something to remember when working with docker images. The image associated with a particular sha hash won't change.

alexbarrera replied:
I guess I trust the tag to represent some intended stationary state of the image, which I guess it might change, but that may indeed be desirable (e.g. to fix a security vulnerability in the base image). Generally, I find the tag useful to know which version/release represents

alexbarrera commented on 01_mpileups/filter_unknown_ref_base/filter_unknown_ref_base.cwl:
```diff
@@ -1,8 +1,14 @@
+
+hints:
+    DockerRequirement:
+      dockerPull: alpine@sha256:06b5d462c92fc39303e6363c65e074559f8d6b1363250027ed5053557e3398c5
```
Any chance we can use a tagged image (for example `alpine:20220715`).

alexbarrera commented on sample_bird_workflow.yml:
```diff
@@ -1,13 +1,13 @@
   - Output1
-alignments_array:
+alignment_sets:
   - { class: File, path: "/data/reddylab/gjohnson/scz/ORI_library_files_and_data/alignment_and_processing/alignments/Input1/Input1_lengthFiltered.bam" }
   - { class: File, path: "/data/reddylab/gjohnson/scz/ORI_library_files_and_data/alignment_and_processing/alignments/Output1/Output1.bam" }
```
I was hoping we could use the sampled files instead of these multi GB ones for the sample `yml` file (since otherwise it becomes a bit hard to test and debug this). I will open an issue about this and we can work on it later.

alexbarrera reviewed (commented):
Other than changing the Docker image, I think it is all good. Q: have you tested this using Docker images of the required software?

tncowart committed:
Fixes some docker issues

Changes alpine to use a tag instead of a sha hash. The image associated with
a tag can change, which can be undesirable. In this case, it's probably fine,
but it's something to remember when working with docker image tags.

Changed some indentation

Added InitialWorkDirRequirement to samtools

samtools tries to write a FASTA index to the same directory the FASTA file is in.
Inside a container this might be read-only. So by adding the InitialWorkDirRequirement
we make sure the FASTA file is mounted in a writeable directory.

alexbarrera committed:
Add missing index .fai file for reference genome as secondaryFiles

tncowart committed:
Remove InitialWorkDirectory requirement

This isn't needed since the FASTA index was added as a
secondary file of the FASTA file

alexbarrera reviewed (approved):
Looks good!
